dosbox-staging/dosbox-staging
# Rework shader loading

# Description

This PR changes the way shader programs and related settings are stored. Instead of keeping just one shader program and recompiling it each time a shader change is requested, every shader is now loaded and compiled once on demand, then stored in a `std::map`. This allows for much quicker transitions between shaders in `crt-auto` sets, and shader loading process in general was reworked a bit to be slightly more sane.

Partially extracted from #2450. The first two commits are self-explanatory high-precision chainsaw surgery, mostly unrelated.

Unfortunately, the work is unfinished (see the next section), and I don't really have a feeling for where it should move. The real reason behind the changes was to have some sane way of loading shaders outside of OpenGL rendering initialization in #2450, and initially the changes were meant to be a part of that PR. The entanglement of `sdlmain.cpp`, `render.cpp`, and `shader_manager.cpp` did me. Couldn't sit on it, so pushing as is. It's up for grabs, if anyone wants it. Sorry.


# Manual testing

- Running with any shader set in the configuration - works;
- Triggering automatic shader switch by resizing the window - works, the storage is populated as expected, but there's still a noticeable hiccup whenever the target shader changes pixel doubling setting;
- Triggering manual shader reload (Ctrl+F2) - fails, not implemented;
- Attempting to load a non-existing shader source file - works, shader manager catches that and instead passes a fallback source (note that it will still attempt to load other sources from the auto set and will retain integer scaling setting when automatic switch is triggered);
- Attempting to load a malformed shader source - fails, either keeps the last successfully compiled shader or outputs nothing.


# Checklist

I have:
